src/internal/analytics/__tests__/hooks/use-funnel-sub-step.test.tsx

```diff
@@ -0,0 +1,157 @@
+import FunnelMetrics from '../../../analytics';
+
+jest.mock('../../../analytics', () => ({
+  __esModule: true,
+  default: {
+    funnelSubStepStart: jest.fn(),
+    funnelSubStepComplete: jest.fn(),
+  },
+}));
+
+describe('useFunnelSubStep hook', () => {
+  test('returns the correct initial context value', () => {
+    const wrapper: React.FC = ({ children }) => (
+      <FunnelSubStepContext.Provider
+        value={{
+          funnelInteractionId: undefined,
+          subStepId: '',
+          stepNumber: 0,
+          stepNameSelector: '',
+          subStepSelector: '',
+          subStepNameSelector: '',
+        }}
+      >
+        {children}
+      </FunnelSubStepContext.Provider>
+    );
+
+    const { result } = renderHook(() => useFunnelSubStep(), { wrapper });
+
+    expect(result.current.stepNumber).toBe(0);
+    expect(result.current.stepNameSelector).toBe('');
+    expect(result.current.subStepSelector).toBe('');
+    expect(result.current.subStepNameSelector).toBe('');
+  });
+
+  test('updates context values', () => {
+    const funnelInteractionId = '123';
+    const stepNumber = 1;
+    const stepNameSelector = 'step1';
+    const subStepSelector = 'subStep1';
+    const subStepNameSelector = 'subStepName1';
+
+    const wrapper: React.FC = ({ children }) => (
+      <FunnelSubStepContext.Provider
+        value={{
+          subStepId: '',
+          funnelInteractionId,
+          stepNumber,
+          stepNameSelector,
+          subStepSelector,
+          subStepNameSelector,
+        }}
+      >
+        {children}
+      </FunnelSubStepContext.Provider>
+    );
+
+    const { result } = renderHook(() => useFunnelSubStep(), { wrapper });
+
+    expect(result.current.funnelInteractionId).toBe(funnelInteractionId);
+    expect(result.current.stepNumber).toBe(stepNumber);
+    expect(result.current.stepNameSelector).toBe(stepNameSelector);
+    expect(result.current.subStepSelector).toBe(subStepSelector);
+    expect(result.current.subStepNameSelector).toBe(subStepNameSelector);
+  });
+
+  test('calls funnelSubStepStart when the substep is focused', () => {
+    const funnelInteractionId = '123';
+    const subStepId = '456';
+    const stepNumber = 1;
+    const stepNameSelector = 'step1';
+    const subStepSelector = 'subStep1';
+    const subStepNameSelector = 'subStepName1';
+
+    const wrapper: React.FC = ({ children }) => (
+      <FunnelSubStepContext.Provider
+        value={{
+          subStepId,
+          funnelInteractionId,
+          stepNumber,
+          stepNameSelector,
+          subStepSelector,
+          subStepNameSelector,
+        }}
+      >
+        {children}
+      </FunnelSubStepContext.Provider>
+    );
+
+    const { result } = renderHook(() => useFunnelSubStep(), { wrapper });
+
+    // Set subStepRef
+    const subStep = document.createElement('div');
+    result.current.subStepRef.current = subStep;
+    result.current.funnelSubStepProps.onFocus({});
```

Comment:
This would look much better if testing with React components instead of direct functions call

Reply:
> I checked the tech design doc, but it mentions nothing about `Analytics*` components and `useFunnel*` hooks.
> 
> So in this PR I also sort of reviewed this components APIs (because this is what all our team is expected to use to instrument their components)

Yes, sorry the core design stayed the same to use React context to maintain the additional state but the implementation details changed due to feedback from the review. The feedback was that the initial generic approach would an additional unnecessary layer we need to maintain and may never use (pre optimization). The doc does not accurately reflect that shift.

Reply:
The focus/blur events were difficult to test. Trying to test them with components runs into jsdom's limitations with focus/blur and they would not fire correctly or at all when trying to focus on an element and tab in and out. The current tests were the best compromise I could find that seemed to work. 

I also wasn't sure how to test this in integ since the dev page has no way to change the stubbed implementation which does nothing so there is nothing to assert as you tab through the fields.

I'm open to better suggestions for a better approach.

Reply:
Converted all tests away from renderHook utility to use React components.